File: tools/objects/source-authoring/lucy-targets/author.mjs
#!/usr/bin/env node
import { readFile } from 'node:fs/promises';
import { resolve } from 'node:path';
import { createHash } from 'node:crypto';

function verify(bytes, entry, label) {
  if (!entry || bytes.length !== entry.expectedBytes ||
      createHash('sha256').update(bytes).digest('hex') !== entry.expectedSha256) {
    throw new Error(`${label} differs from its source manifest pin.`);
  }
}

async function main() {
  if (process.argv.length !== 3) throw new Error('Usage: node author.mjs <body-source-directory>');
  const source = resolve(process.argv[2]);
  const manifest = JSON.parse(await readFile(resolve(source, 'manifest.json')));
  const measurementBytes = await readFile(resolve(source, 'measurements.json'));
  verify(measurementBytes, manifest.documents.find(entry => entry.path === 'measurements.json'), 'Measurement file');
  const measurements = JSON.parse(measurementBytes);
  const axes = measurements.constraints?.fullAxesKm;
  if (measurements.schema !== 'cssearth-approximate-ellipsoid@1' || !Array.isArray(axes) ||
      axes.length !== 3 || !axes.every(value => Number.isFinite(value) && value > 0)) {
    throw new Error('Expected the existing approximate-ellipsoid measurements and three positive full axes.');
  }
  const [a, c2, c] = axes.map(x => x / 2);
  const shape = [];
  for (let lat = -90; lat <= 90; lat += 5) for (let lon = 0; lon <= 360; lon += 5) {
    const p = lat * Math.PI / 180, l = lon * Math.PI / 180;
    const r = 1 / Math.sqrt((Math.cos(p) * Math.cos(l) / a) ** 2 + (Math.cos(p) * Math.sin(l) / c2) ** 2 + (Math.sin(p) / c) ** 2);
    shape.push(`${lon} ${lat} ${r.toFixed(12)}`);
  }
  const table = Buffer.from(shape.join('\n') + '\n');
  verify(table, manifest.inputs.find(entry => entry.path === 'shape/ellipsoid.tab'), 'Ellipsoid table');
  process.stdout.write(table);
}

main().catch(error => { console.error(error.message); process.exitCode = 1; });

File: tools/objects/source-authoring/trans-neptunian/author.py
#!/usr/bin/env python3
from pathlib import Path
import hashlib
import json
import math
import sys


def verify(data, entry, label):
    if entry is None:
        raise ValueError(f'{label} differs from its source manifest pin.')


def main():
    if len(sys.argv) != 2:
        raise ValueError('Usage: python3 author.py <body-source-directory>')
    source = Path(sys.argv[1])
    manifest = json.loads((source / 'manifest.json').read_bytes())
    measurement_bytes = (source / 'measurements.json').read_bytes()
    verify(measurement_bytes, next((entry for entry in manifest['documents']
                                   if entry['path'] == 'measurements.json'), None), 'Measurement file')
    measurements = json.loads(measurement_bytes)
    full_axes = measurements.get('fullAxesKm')
    if (measurements.get('schema') != 'cssearth-trans-neptunian-source@1' or
            not isinstance(full_axes, list) or len(full_axes) != 3 or
            not all(isinstance(value, (int, float)) and math.isfinite(value) and value > 0
                    for value in full_axes)):
        raise ValueError('Expected the existing trans-Neptunian measurements and three positive full axes.')
    axes = [n / 2 for n in full_axes]
    lines = []
    for lat in range(-90, 91, 5):
        for lon in range(0, 361, 5):
            p, l = math.radians(lat), math.radians(lon)
            r = 1 / math.sqrt((math.cos(p)*math.cos(l)/axes[0])**2 + (math.cos(p)*math.sin(l)/axes[1])**2 + (math.sin(p)/axes[2])**2)
            lines.append(f'{lon} {lat} {r:.12f}')
    table = ('\n'.join(lines) + '\n').encode()
    verify(table, next((entry for entry in manifest['inputs']
                        if entry['path'] == 'shape/ellipsoid.tab'), None), 'Ellipsoid table')
    sys.stdout.buffer.write(table)


if __name__ == '__main__':
    try:
        main()
    except (ValueError, KeyError, OSError) as error:
        print(error, file=sys.stderr)
        sys.exit(1)
